src/mysql_conn.erl

```diff
@@ -91,54 +92,109 @@ init(Opts) ->
         N when N > 0 -> N
     end,
 
+    State0 = #state{
+        tcp_opts = TcpOpts,
+        ssl_opts = SSLOpts,
+        host = Host, port = Port,
+        user = User, password = Password,
+        database = Database,
+        queries = Queries, prepares = Prepares,
+        log_warnings = LogWarn,
+        connect_timeout = ConnectTimeout,
+        ping_timeout = PingTimeout,
+        query_timeout = QueryTimeout,
+        query_cache_time = QueryCacheTime,
+        cap_found_rows = (SetFoundRows =:= true)
+    },
+
+    case proplists:get_value(connect_mode, Opts, synchronous) of
+        synchronous ->
+            case do_connect(State0) of
+                {ok, State1} ->
+                    {ok, State1};
+                {error, Reason} ->
+                    {stop, Reason}
+            end;
+        asynchronous ->
+            gen_server:cast(self(), connect),
+            {ok, State0};
+        lazy ->
+            {ok, State0}
+    end.
+
+do_connect(#state{connect_timeout = ConnectTimeout} = State) ->
+    Self = self(),
+    Pid = spawn_link(
+        fun () ->
+            case do_connect1(State) of
+                {ok, #state{sockmod = SockMod, socket = Socket} = State1} ->
+                    SockMod:controlling_process(Socket, Self),
+                    Self ! {self(), {ok, State1}};
+                {error, _} = E ->
+                    Self ! {self(), E}
+            end
+        end
+    ),
+    receive
+        {Pid, {ok, State1}} ->
+            do_post_connect(State1);
+        {Pid, {error, _} = E} ->
+            E
+    after ConnectTimeout ->
+        unlink(Pid),
+        exit(Pid, kill),
+        {error, timeout}
+    end.
+
+do_connect1(#state{tcp_opts = TcpOpts, host = Host, port = Port} = State) ->
     %% Connect socket
     SockOpts = [binary, {packet, raw}, {active, false}, {nodelay, true}
                 | TcpOpts],
-    {ok, Socket0} = SockMod0:connect(Host, Port, SockOpts),
+    {ok, Socket} = gen_tcp:connect(Host, Port, SockOpts),
```

Comment:
Here `gen_tcp:connect(Address, Port, Options, Timeout)` can be used, instead of the extra process for timeout handling.